packages/runtime/src/state/react-builder-preview.ts

```diff
@@ -237,27 +237,33 @@ function startHandlingPointerMoveEvent(): ThunkAction<() => void, State, unknown
   }
 }
 
-function startHandlingFocusEvents(): () => void {
-  window.addEventListener('focusin', handleFocusIn)
-  window.addEventListener('focusout', handleFocusOut)
+function startHandlingFocusEvents(): ThunkAction<() => void, State, unknown, Action> {
+  return (_dispatch, getState) => {
+    window.addEventListener('focusin', handleFocusIn)
+    window.addEventListener('focusout', handleFocusOut)
 
-  return () => {
-    window.removeEventListener('focusin', handleFocusIn)
-    window.removeEventListener('focusout', handleFocusOut)
-  }
+    return () => {
+      window.removeEventListener('focusin', handleFocusIn)
+      window.removeEventListener('focusout', handleFocusOut)
+    }
+
+    function handleFocusIn(event: FocusEvent) {
+      if (ReactPage.getEditMode(getState()) === EditMode.EditModes.INTERACT) return
 
-  function handleFocusIn(event: FocusEvent) {
-    if (!(event.target instanceof window.HTMLElement) || !event.target.isContentEditable) {
-      window.parent.focus()
+      if (!(event.target instanceof window.HTMLElement) || !event.target.isContentEditable) {
+        window.parent.focus()
+      }
     }
-  }
 
-  function handleFocusOut(event: FocusEvent) {
-    if (
-      !(event.relatedTarget instanceof window.HTMLElement) ||
-      !event.relatedTarget.isContentEditable
-    ) {
-      window.parent.focus()
+    function handleFocusOut(event: FocusEvent) {
+      if (ReactPage.getEditMode(getState()) === EditMode.EditModes.INTERACT) return
+
+      if (
+        !(event.relatedTarget instanceof window.HTMLElement) ||
+        !event.relatedTarget.isContentEditable
+      ) {
+        window.parent.focus()
+      }
```

Comment:
Right now, the reason we can capture the blur when in content mode when pressing `esc` is because we're attaching `esc` listener in the Text component:
```ts
if (isHotkey('escape')(event)) {
  lastController.current?.blur()

  return true
}
```

When the Text component blurs, this event listener will catch the `focusout`, and then focus the `window.parent`.

We cannot do that to any other components in interaction mode without attaching the `esc` event listener. Pressing `esc` won't trigger blur of any HTML element, so we cannot capture the `focusout` event.

Reply:
I would prefer if we didn't hard-code these assumptions into our components like this. We should discuss how we want to handle entering and exiting interaction mode without needing to add an `esc` listener

Reply:
Looks like we need a keyboard listener if we want to make this work. The reason focus worked before is that for content mode and `contenteditable`, hitting <kbd>Esc</kbd> does fire a `focusout` event, but with other focusable elements, like buttons, hitting <kbd>Esc</kbd> does not trigger the `focusout` event.